mldsa/fips202/fips202.c

```diff
@@ -394,15 +394,26 @@ static void keccak_init(uint64_t s[25])
  *
  * Returns new position pos in current block
  **************************************************/
-static unsigned int keccak_absorb(uint64_t s[25], unsigned int pos,
-                                  unsigned int r, const uint8_t *in,
-                                  size_t inlen)
+static unsigned int keccak_absorb(uint64_t s[MLD_KECCAK_LANES],
+                                  unsigned int pos, unsigned int r,
+                                  const uint8_t *in, size_t inlen)
+__contract__(
+  requires(r >= 0 && r < sizeof(uint64_t) * MLD_KECCAK_LANES)
+  requires(pos < r && pos >= 0)
+  requires(memory_no_alias(s, sizeof(uint64_t) * MLD_KECCAK_LANES))
+  requires(memory_no_alias(in, inlen))
+  assigns(memory_slice(s, sizeof(uint64_t) * MLD_KECCAK_LANES)))
 {
   unsigned int i;
 
-  while (pos + inlen >= r)
+  if (pos + inlen >= r)
```

Comment:
I don't think the removal of the `while` loop is necessary. You can use the same invariant as you do below, just add `invariant(pos < r)`

Reply:
@manastasova I took the liberty to adjust this in a follow-up commit.